src/clients/dotnet/TigerBeetle/Request.cs

```diff
@@ -1,151 +1,143 @@
 using System;
+using System.Buffers;
 using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
 using static TigerBeetle.AssertionException;
 
 namespace TigerBeetle;
 
-internal struct Packet
+
+internal abstract class NativeRequest
 {
-    public readonly unsafe TBPacket* Pointer;
+    private Tuple<IMemoryOwner<TBPacket>, MemoryHandle>? packetMemory = null;
 
-    public unsafe Packet(TBPacket* pointer)
+    protected unsafe void Submit(NativeClient nativeClient, TBOperation operation, void* data, int len)
     {
-        Pointer = pointer;
+        AssertTrue(packetMemory == null);
```

Comment:
Believe it should have at least one assert for `var request = GCHandle.FromIntPtr(packet->userData).Target ?? throw Exception`. A handle with a null Target seems to only happen for [Weak handles](<https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandletype?view=net-8.0>) (otherwise its kept alive until Free) or for [already freed/invalid handles](<https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle.target?view=net-8.0#exceptions>)

Reply:
> Believe it should have at least one assert 

I agree, but I'm afraid the exception won't be caught by the caller, and possibly will only prevent the callback to complete...

In fact it should panic instead!
Maybe using Environment.FailFast:

http://msdn.microsoft.com/en-us/library/ms131100%28v=vs.110%29.aspx